### PR TITLE
Log which requirements were or weren't satisfied

### PR DESCRIPTION
**Pull request by konstin**
Previously, a noop `uv pip install` would only show "Audited {} package(s)" but no details, not even with `-vv`. Now it debug logs which requirements were met and it also debug logs which requirement was missing to trigger the full routine, allowing it investigate caching behaviour.

First `uv pip install -v jupyter`:

```
DEBUG At least one requirement is not satisfied: jupyter
```

Second `uv pip install -v jupyter`:

```
DEBUG Found a virtualenv named .venv at: /home/konsti/projects/uv-main/.venv
DEBUG Cached interpreter info for Python 3.12.1, skipping probing: .venv/bin/python
DEBUG Using Python 3.12.1 environment at .venv/bin/python
DEBUG Trying to lock if free: .venv/.lock
DEBUG Requirement satisfied: anyio
DEBUG Requirement satisfied: anyio>=3.1.0
DEBUG Requirement satisfied: argon2-cffi-bindings
DEBUG Requirement satisfied: argon2-cffi>=21.1
DEBUG Requirement satisfied: arrow>=0.15.0
DEBUG Requirement satisfied: asttokens>=2.1.0
DEBUG Requirement satisfied: async-lru>=1.0.0
DEBUG Requirement satisfied: attrs>=22.2.0
DEBUG Requirement satisfied: babel>=2.10
...
DEBUG Requirement satisfied: webencodings
DEBUG Requirement satisfied: webencodings>=0.4
DEBUG Requirement satisfied: websocket-client>=1.7
DEBUG Requirement satisfied: widgetsnbextension~=4.0.10
DEBUG All editables satisfied: 
Audited 1 package in 12ms
```

This will clash with the `tool.uv.sources` PR, i'll rebase it on top.

### CODE REVIEW

File: crates/uv-installer/src/site_packages.rs [hunk 1]
```diff
@@ -495,15 +498,24 @@ impl<'a> SitePackages<'a> {
...
         }
 
-        Ok(true)
+        Ok(SatisfiesResult::Fresh {
+            recursive_requirements: seen,
+        })
     }
 }
 
+pub enum SatisfiesResult {
```
Comment (zanieb) [inline crates/uv-installer/src/site_packages.rs:512]: Should this derive debug and such?

File: crates/uv-installer/src/site_packages.rs [hunk 2]
```diff
@@ -495,15 +498,24 @@ impl<'a> SitePackages<'a> {
...
+    Fresh {
+        recursive_requirements: FxHashSet<RequirementEntry>,
+    },
+    Unsatisfied(String),
```
Comment (zanieb) [inline crates/uv-installer/src/site_packages.rs:516]: Can you add a comment as to why we only have one item here?

File: crates/uv/src/commands/pip_install.rs [hunk 3]
```diff
@@ -173,28 +175,46 @@ pub(crate) async fn pip_install(
...
+            SatisfiesResult::Fresh {
+                recursive_requirements,
+            } => {
+                if enabled!(Level::DEBUG) {
+                    for requirement in recursive_requirements
+                        .iter()
+                        .map(ToString::to_string)
+                        .sorted()
+                    {
+                        debug!("Requirement satisfied: {requirement}");
```
Comment (zanieb) [inline crates/uv/src/commands/pip_install.rs:189]: Can we make these ones trace?

File: crates/uv-installer/src/site_packages.rs [hunk 4]
```diff
@@ -293,13 +293,14 @@ impl<'a> SitePackages<'a> {
         Ok(diagnostics)
     }
 
-    /// Returns `true` if the installed packages satisfy the given requirements.
+    /// Returns `Ok(None)` if the installed packages satisfy the given requirements or the
```
Comment (zanieb) [inline crates/uv-installer/src/site_packages.rs:296]: This looks wrong

**Review by zanieb — APPROVED** [review]

File: crates/uv/src/commands/pip_install.rs (hunk 3 above)
Comment (zanieb) [reply crates/uv/src/commands/pip_install.rs:189]: I guess I'm actually okay with debug.

File: crates/uv-installer/src/site_packages.rs (hunk 1 above)
Comment (konstin) [reply crates/uv-installer/src/site_packages.rs:513]: Added a debug derive and a docstring

File: crates/uv/src/commands/pip_install.rs (hunk 3 above)
Comment (konstin) [reply crates/uv/src/commands/pip_install.rs:189]: These should stay debug messages, they are similar to the message about the packages that we installed. They are the nearly final messages (this is the early exist path), they are only followed by:

```
DEBUG All editables satisfied: [...]
Audited 1 package in 12ms
```

File: crates/uv-installer/src/site_packages.rs (hunk 4 above)
Comment (konstin) [reply crates/uv-installer/src/site_packages.rs]: good catch

**Pull request merged**